test/test_internet.py

```diff
@@ -9,22 +9,40 @@
 import pytest
 import warnings
 from urllib.error import URLError
+import contextlib
 
 import torchvision.datasets.utils as utils
 from common_utils import get_tmp_dir
 
 
-class TestDatasetUtils:
+def patch_url_redirection(mocker, redirect_url):
+    class Response:
+        def __init__(self, url):
+            self.url = url
+
+    @contextlib.contextmanager
+    def patched_opener(*args, **kwargs):
+        yield Response(redirect_url)
+
+    mocker.patch("torchvision.datasets.utils.urllib.request.urlopen", new=patched_opener)
+
 
-    def test_get_redirect_url(self):
+class TestDatasetUtils:
+    def test_get_redirect_url(self, mocker):
         url = "http://www.vision.caltech.edu/visipedia-data/CUB-200-2011/CUB_200_2011.tgz"
-        expected = "https://drive.google.com/file/d/1hbzc_P1FuxMkcabkgn9ZKinBwW683j45/view"
+        expected_redirect_url = "https://drive.google.com/file/d/1hbzc_P1FuxMkcabkgn9ZKinBwW683j45/view"
+
+        patch_url_redirection(mocker, expected_redirect_url)
```

Comment:
If I understand the logic correctly, should we now ensure that the mocker/urlopen was called exactly twice, once with the first URL and once with the second URL?

Also, since we don't need to use real URLs anymore, perhaps we could just create obviously fake ones like `url = "some_url"` and `expected_redirect_url = "some_other_url"`?

Reply:
> If I understand the logic correctly, should we now ensure that the mocker/urlopen was called exactly twice, once with the first URL and once with the second URL?

I can add that, yes.

> Also, since we don't need to use real URLs anymore, perhaps we could just create obviously fake ones like `url = "some_url"` and `expected_redirect_url = "some_other_url"`?

Good point.

Reply:
After some more thought, I think we should not test the mocked `urlopen` further.

1. Since the call actually returns a context manager, we would need to create our custom mock instead of relying on the `new` kwarg. AFAIK there is no easy way to do it.
2. We shouldn't really care. As long as `_get_redirect_url` returns the correct URL, we shouldn't really care about the internals.

Reply:
We can do it by replacing `new` with `side_effect`, this will return a MagicMock (with a call_count attribute that we can validate) instead of the `patched_opener`  function:

```patch
@ test/test_datasets_utils.py:32 @ def patch_url_redirection(mocker, redirect_url):
    def patched_opener(*args, **kwargs):
        yield Response(redirect_url)

-    return mocker.patch("torchvision.datasets.utils.urllib.request.urlopen", new=patched_opener)
+    return mocker.patch("torchvision.datasets.utils.urllib.request.urlopen", side_effect=patched_opener)


class TestDatasetsUtils:
@ test/test_datasets_utils.py:40 @ class TestDatasetsUtils:
        url = "https://url.org"
        expected_redirect_url = "https://redirect.url.org"

-        patch_url_redirection(mocker, expected_redirect_url)
+        mocked = patch_url_redirection(mocker, expected_redirect_url)

        actual = utils._get_redirect_url(url)
+      assert mocked.call_count == 2
        assert actual == expected_redirect_url
```

Regarding 2: I would agree that we shouldn't have to care about the internals, but I feel like we're already concerned with internals by creating a Response object with a url attribute. So we might as well fully validate the logic, IMHO.

Anyway, no strong opinion, we can also merge as-is :)